### PR TITLE
feat(rbg): support coordinated segment scheduling

### PR DESCRIPTION
**Pull request by Syspretor**
# Segment Scheduling for Multi-Level Gang Scheduling

## Problem Statement

In distributed AI serving scenarios like Prefill-Decode (PD) separation, we need multi-level gang scheduling:

1. **Instance-Level Gang**: Each RoleInstance needs to be a PodGroup to ensure multi-node deployment instances run properly
2. **Role-Level Gang**: At the Role level, we also need a PodGroup to maintain the correct ratio between Prefill and Decode roles

**Challenge**: In resource-constrained environments, without proper coordination, the cluster might successfully deploy only Prefills or only Decodes, leading to an unusable system. We need to ensure that both roles are deployed proportionally according to their segment configurations.

**Solution**: We introduce **Segment Scheduling** - a coordination strategy that deploys replicas in segments (batches) until reaching the desired final count, ensuring coordinated progression across all roles.

### 1. New `SegmentScheduling` Strategy

Added to `CoordinationStrategy` in the RoleBasedGroup CRD:

```go
type CoordinationStrategy struct {
    // ... existing fields ...
    
    // SegmentScheduling enables segment-based deployment strategy
    // +optional
    SegmentScheduling *SegmentScheduling `json:"segmentScheduling,omitempty"`
}

type SegmentScheduling struct {
    // SegmentSize defines the deployment segments and their replica counts for each role.
    // Key is the role name, value is the number of replicas in this segment.
    SegmentSize map[string]int32 `json:"segment"`
    
    // PartitionStrategy defines the strategy for deploying segments in a partitioned manner.
    // Determines whether to wait for the current segment to be ready before creating the next one.
    // Defaults to SequentialWait.
    // +optional
    // +kubebuilder:validation:Enum={Sequential,SequentialWait}
    // +kubebuilder:default=SequentialWait
    PartitionStrategy SegmentPartitionStrategyType `json:"partitionStrategy,omitempty"`
}

type SegmentPartitionStrategyType string

const (
    // SegmentPartitionStrategySequential creates segments one by one without waiting for readiness.
    SegmentPartitionStrategySequential SegmentPartitionStrategyType = "Sequential"
    
    // SegmentPartitionStrategySequentialWait creates segments one by one and waits for each to be ready.
    SegmentPartitionStrategySequentialWait SegmentPartitionStrategyType = "SequentialWait"
)
```

### 2. Coordination Requirements Structure

```go
type Coordination struct {
    Strategy *CoordinationStrategy `json:"strategy,omitempty"`
}

type CoordinationRequirement struct {
    // CoordinationRequirements defines multiple coordination strategies
    CoordinationRequirements []Coordination `json:"coordinationRequirements,omitempty"`
}
```

### Basic Example: PD Separation with Segment Scheduling

```yaml
apiVersion: workloads.instai.io/v1alpha1
kind: RoleBasedGroup
metadata:
  name: llm-service
spec:
  roles:
  - name: prefill
    replicas: 100
    # ... workload spec ...
  
  - name: decode
    replicas: 50
    # ... workload spec ...
  
  coordinationRequirements:
  - strategy:
      segmentScheduling:
        segment:
          prefill: 10  # Deploy 10 prefill replicas per segment
          decode: 5    # Deploy 5 decode replicas per segment
        partitionStrategy: SequentialWait
```

**Deployment Flow**:
1. **Segment 1**: Deploy 10 Prefills + 5 Decodes → Wait until all ready
2. **Segment 2**: Deploy next 10 Prefills + 5 Decodes → Wait until all ready
3. **Segment 3**: Deploy next 10 Prefills + 5 Decodes → Wait until all ready
4. ... Continue until reaching 100 Prefills + 50 Decodes

### Advanced Example: Multi-Coordination with Different Roles

```yaml
apiVersion: workloads.instai.io/v1alpha1
kind: RoleBasedGroup
metadata:
  name: complex-llm-service
spec:
  roles:
  - name: prefill
    replicas: 300
  
  - name: decode
    replicas: 100
  
  - name: router
    replicas: 50
  
  coordinationRequirements:
  # Coordination 1: Prefill and Decode must scale together
  - strategy:
      segmentScheduling:
        segment:
          prefill: 5
          decode: 3
        partitionStrategy: SequentialWait
  
  # Coordination 2: Decode and Router must scale together
  - strategy:
      segmentScheduling:
        segment:
          decode: 3
          router: 2
        partitionStrategy: SequentialWait
```

**Key Points**:
- `decode` appears in both coordinations with the same segment size (3) - this is required
- Coordinations block each other: if Coordination 1 is not ready, Coordination 2 will also wait
- Ensures all three roles scale proportionally

#### SequentialWait (Default - Recommended)
```yaml
segmentScheduling:
  segment:
    prefill: 10
    decode: 5
  partitionStrategy: SequentialWait
```

- Deploys one segment at a time
- **Waits** for all replicas in current segment to be ready before advancing
- **Use case**: Production environments where stability is critical

#### Sequential (Aggressive)
```yaml
segmentScheduling:
  segment:
    prefill: 10
    decode: 5
  partitionStrategy: Sequential
```

- Deploys segments continuously without waiting
- Does NOT wait for readiness between segments
- **Use case**: Fast deployment scenarios or testing

### 1. Segment Calculation

The scheduler calculates the minimum number of complete segments deployed across all roles:

```
minFullSegment = min(prefill_replicas / prefill_segment_size, decode_replicas / decode_segment_size)
```

**Example**:
- Prefill: 25 replicas deployed, segment size = 10 → 2 complete segments
- Decode: 18 replicas deployed, segment size = 5 → 3 complete segments
- **minFullSegment = 2**

### 2. Readiness Check

For `SequentialWait` strategy, checks if the current segment is fully ready:

```go
isSegmentReady(segmentIndex) {
    for each role {
        expectedReplicas = segmentIndex * segmentSize
        if currentReplicas < expectedReplicas OR readyReplicas < expectedReplicas {
            return false
        }
    }
    return true
}
```

### 3. Target Calculation

```go
if partitionStrategy == SequentialWait && !isSegmentReady(minFullSegment) {
    // Keep current targets (wait)
    return currentReplicas
}

// Advance to next segment
return (minFullSegment + 1) * segmentSize
```

### 4. Multi-Coordination Merge Logic

When multiple coordinations exist:

1. Calculate target replicas for each coordination independently
2. Check if each coordination can progress (not blocked by readiness)
3. For shared roles (appearing in multiple coordinations):
   - Take the **minimum** target across all coordinations
   - Ensures coordinated progression
4. If any role in a coordination is blocked, ALL roles in that coordination are blocked

## Conflict Detection

The system detects and rejects conflicting configurations:

### Segment Size Conflict
```yaml
coordinationRequirements:
- strategy:
    segmentScheduling:
      segment:
        prefill: 10  # ❌ Conflict!
        
- strategy:
    segmentScheduling:
      segment:
        prefill: 5   # ❌ Same role, different size
```

**Error**: `segment size conflict for role "prefill"`

### Partition Strategy Conflict
```yaml
coordinationRequirements:
- strategy:
    segmentScheduling:
      segment:
        prefill: 10
      partitionStrategy: SequentialWait  # ❌ Conflict!
        
- strategy:
    segmentScheduling:
      segment:
        prefill: 10
      partitionStrategy: Sequential      # ❌ Same role, different strategy
```

**Error**: `partition strategy conflict for role "prefill"`

### ✅ Proportional Scaling
- Maintains correct ratios between roles during deployment
- Prevents partial deployments (e.g., only Prefills without Decodes)

### ✅ Resource-Aware
- Respects desired replica limits
- Caps targets at `min(segment_target, desired_replicas)`

### ✅ Multi-Coordination Support
- Multiple coordination requirements can coexist
- Automatic blocking and synchronization between coordinations

### ✅ Flexible Strategies
- `SequentialWait`: Safe, production-ready (waits for readiness)
- `Sequential`: Fast deployment (no waiting)

### ✅ Conflict Detection
- Validates segment sizes across coordinations
- Validates partition strategies across coordinations
- Fails fast with clear error messages

### Core Components

1. **SegmentScheduler** (`pkg/coordination/segmentscheduling/scheduler.go`)
   - `NewScheduler(rbg, roleStatuses)`: Creates scheduler for an RBG
   - `CalculateTargetReplicas()`: Main entry point, returns target replicas per role
   - `calculateMinFullSegment()`: Calculates minimum complete segments
   - `isSegmentReady()`: Checks if a segment is fully ready
   - `getSegmentTargets()`: Calculates target replicas for a segment
   - `detectSegmentConflicts()`: Validates configuration conflicts
   - `mergeSegmentTargets()`: Merges targets from multiple coordinations

2. **Integration** (`internal/controller/workloads/rolebasedgroup_controller.go`)
   ```go
   if segmentscheduling.IsActive(rbg) {
       scheduler := segmentscheduling.NewScheduler(rbg, roleStatuses)
       targetReplicas, err := scheduler.CalculateTargetReplicas()
       // Apply target replicas to roles
   }
   ```

### Testing

Comprehensive test coverage includes:
- ✅ Single coordination scenarios
- ✅ Multi-coordination scenarios with overlapping roles
- ✅ Conflict detection (segment size and partition strategy)
- ✅ Sequential vs SequentialWait strategies
- ✅ Partial segment handling
- ✅ Desired replicas limits
- ✅ Initial deployment (0 replicas)
- ✅ Readiness blocking behavior
- ✅ Cross-coordination blocking

### Before (Manual Coordination)
```yaml
# Users had to manually manage scaling
kubectl scale rbg llm-service --replicas=10:5  # Step 1
# Wait...
kubectl scale rbg llm-service --replicas=20:10 # Step 2
# Wait...
```

### After (Automated Segment Scheduling)
```yaml
apiVersion: workloads.instai.io/v1alpha1
kind: RoleBasedGroup
metadata:
  name: llm-service
spec:
  roles:
  - name: prefill
    replicas: 100
  - name: decode
    replicas: 50
  
  coordinationRequirements:
  - strategy:
      segmentScheduling:
        segment:
          prefill: 10
          decode: 5
        partitionStrategy: SequentialWait
```

The controller automatically handles incremental scaling!

## Best Practices

1. **Choose Appropriate Segment Sizes**
   - Too small: Many segments, slower overall deployment
   - Too large: Wastes resources if deployment fails mid-segment
   - Recommended: 5-10% of total desired replicas per segment

2. **Use SequentialWait in Production**
   - Ensures stability between segments
   - Prevents cascading failures

3. **Maintain Consistent Ratios**
   - Segment sizes should maintain your desired role ratios
   - Example: 2:1 Prefill:Decode ratio → use 10:5 or 20:10 segments

4. **Monitor Deployment Progress**
   - Check RoleBasedGroup status for current replica counts
   - Observe segment progression through events

5. **Avoid Conflicts**
   - If a role appears in multiple coordinations, use the same segment size
   - Use the same partition strategy across all coordinations for a role

## Future Enhancements

- [ ] Topology-aware segment scheduling based on network domains or GPU topology affinity
- [ ] Progress tracking and metrics

## References

- [Examples](examples/segment-scheduling/)

### CODE REVIEW

File: api/workloads/v1alpha1/rolebasedgroup_types.go [hunk 1]
```diff
@@ -62,8 +62,39 @@ type Coordination struct {
...
+
+	// SegmentScheduling defines the coordination strategies for segment-based deployment.
+	// It enables deploying workloads in segments (batches) with configurable partition strategies.
+	SegmentScheduling *SegmentScheduling `json:"segmentScheduling,omitempty"`
+}
+
+type SegmentScheduling struct {
+	// Segment defines the deployment segments and their replica counts for each role.
+	// Key is the role name, value is the number of replicas in this segment.
+	Segment map[string]int `json:"segment"`
```
Comment (cheyang) [inline api/workloads/v1alpha1/rolebasedgroup_types.go]: What about `segmentSize` ?

File: pkg/reconciler/segment_scheduler.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	}
+	return 0
+}
+
+// isCurrentSegmentReady checks if the currently deployed segment is ready across all roles
+// A segment is ready when ALL roles have their current replicas fully ready
+// This ensures coordinated progression across all roles in a segment
+func (s *SegmentScheduler) isCurrentSegmentReady(
+	roleStatusMap map[string]workloadsv1alpha1.RoleStatus,
+	segmentConfig map[string]int,
```
Comment (cheyang) [inline pkg/reconciler/segment_scheduler.go]: How about using `map[string]int32` to keep consistency?

File: pkg/reconciler/segment_scheduler.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+		currentReplicas := int32(0)
+		if status, ok := roleStatusMap[roleName]; ok {
+			currentReplicas = status.Replicas
+		}
+
+		// Calculate how many segments have been deployed for this role
+		deployedSegments := int(currentReplicas) / segmentSize
+		if currentReplicas%int32(segmentSize) != 0 {
+			// Partial segment counts as deployed but not complete
+			deployedSegments++
```
Comment (cheyang) [inline pkg/reconciler/segment_scheduler.go]: This counts partial segments as "deployed," leading to incorrect minimum segment calculations. Example: segmentSize=4, currentReplicas=3 → deployedSegments=1, but the system expects 4 ready replicas before moving to the next segment.

I suggest calculating  `minCompleteSegments`.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/reconciler/segment_scheduler.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	// Get partition strategy, default to SequentialWait
+	partitionStrategy := segmentScheduling.PartitionStrategy
+	if partitionStrategy == "" {
+		partitionStrategy = workloadsv1alpha1.SegmentPartitionStrategySequentialWait
+	}
+
+	targetReplicas := make(map[string]int32)
+	roleStatusMap := s.buildRoleStatusMap()
+
+	// For SequentialWait strategy, check if current segment is ready before proceeding
```
Comment (cheyang) [inline pkg/reconciler/segment_scheduler.go]: How about making the logic easy to understand by 

```golang
func (s *SegmentScheduler) CalculateTargetReplicas() (map[string]int32, error) {
// ...
// Calculate the minimum completed full segment for all roles
minFullSegment := s.calculateMinFullSegment(roleStatusMap, segmentConfig)

// Check if this segment is fully ready
if s.isSegmentReady(minFullSegment, roleStatusMap, segmentConfig) {
// Advance to the next segment
return s.getSegmentTargets(minFullSegment + 1), nil
}

// Keep the current segment
return s.getSegmentTargets(minFullSegment), nil
}
```

**Review by cheyang — CHANGES_REQUESTED** [review]

File: internal/controller/workloads/rolebasedgroup_controller.go [hunk 5]
```diff
@@ -194,7 +194,17 @@ func (r *RoleBasedGroupReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 		r.recorder.Eventf(
 			rbg, corev1.EventTypeWarning, FailedUpdateStatus,
-			"Failed to update status for %s: %v", rbg.Name, err,
+			"Failed to calculate rolling update strategy for %s: %v", rbg.Name, err,
+		)
+		return ctrl.Result{}, err
+	}
+
+	// Calculate segment target replicas if segment scheduling is enabled
+	segmentTargetReplicas, err := r.CalculateSegmentTargetReplicas(rbg, roleStatuses)
```
Comment (cheyang) [inline internal/controller/workloads/rolebasedgroup_controller.go]: The controller calls CalculateSegmentTargetReplicas() in Reconcile(), then updateRBGStatus() recomputes it by iterating all coordinations again. It may lead  performance overhead and potential inconsistency.

How about passing the  segmentScheduler with computed segmentTargetReplicas as parameters to updateRBGStatus():

```golang
f err := r.updateRBGStatus(ctx, rbg, roleStatuses, segmentScheduler); err != nil {
    return ctrl.Result{}, err
```

File: api/workloads/v1alpha1/rolebasedgroup_types.go [hunk 6]
```diff
@@ -62,8 +62,39 @@ type Coordination struct {
...
+	// Key is the role name, value is the number of replicas in this segment.
+	SegmentSize map[string]int32 `json:"segment"`
+
+	// PartitionStrategy defines the strategy for deploying segments in a partitioned manner.
+	// Determines whether to wait for the current segment to be ready before creating the next one.
+	// Defaults to SegmentPartitionStrategySequentialWait.
+	// +optional
+	// +kubebuilder:validation:Enum={Sequential,SequentialWait}
+	// +kubebuilder:default=SequentialWait
+	PartitionStrategy SegmentPartitionStrategyType `json:"partitionStrategy,omitempty"`
```
Comment (cheyang) [inline api/workloads/v1alpha1/rolebasedgroup_types.go:82]: The proposed API changes look excellent. Renaming PartitionStrategy to Progression makes the intent much clearer from a user's perspective, as it describes how the deployment moves forward. WDYT?

```yaml
type SegmentScheduling struct {
	

	// Progression defines how the controller moves from one segment to the next.
	// It determines the timing and conditions for advancing through segments.
	// Defaults to OrderedReady.
	// +optional
	Progression SegmentProgressionType `json:"progression,omitempty"`
}

// SegmentProgressionType defines the progression strategy for segment scheduling.
type SegmentProgressionType string

const (
	// SegmentProgressionOrderedReady ensures strict sequential deployment with readiness checks.
	// Behavior: Deploy Segment 1 -> Wait for Readiness -> Deploy Segment 2.
	// (Formerly: SequentialWait)
	SegmentProgressionOrderedReady SegmentProgressionType = "OrderedReady"

	// SegmentProgressionOrdered ensures sequential creation without waiting.
	// Behavior: Deploy Segment 1 -> Immediately deploy Segment 2.
	// It preserves the order of creation but proceeds to the next segment without blocking.
	// (Formerly: Sequential)
	SegmentProgressionOrdered SegmentProgressionType = "Ordered"

	// SegmentProgressionParallel enables simultaneous deployment of all segments.
	// Behavior: Deploy Segment 1, Segment 2, Segment 3... at the same time.
	// In implementation, this effectively sets the TargetReplicas to the total sum of all segments immediately.
	SegmentProgressionParallel SegmentProgressionType = "Parallel"
)
```

File: pkg/coordination/segmentscheduling/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+// 1. Detecting conflicts where the same role has different segment sizes
+// 2. Merging target replicas from multiple coordinations
+// 3. Ensuring coordinated progression across all coordinations with SequentialWait strategy
+func (s *SegmentScheduler) CalculateTargetReplicas() (map[string]int32, error) {
+	if len(s.segmentSchedulingPolicies) == 0 {
+		return nil, nil
+	}
+
+	// Check for conflicts: same role with different segment sizes
+	if err := s.detectSegmentConflicts(s.segmentSchedulingPolicies); err != nil {
```
Comment (cheyang) [inline pkg/coordination/segmentscheduling/scheduler.go:64]: I recommend adding validation for segmentSize to ensure it is a positive integer and does not exceed the total desired replicas.

**Review by cheyang — COMMENTED** [review]
Thanks for the iteration on this, @Syspretor. The overall design (dedicated `pkg/coordination/segmentscheduling` package, conflict detection, multi-coordination merge) is heading in a good direction. A few things still need attention before this can move forward:

---

**Merge conflicts** -- This branch has conflicts against main and hasn't been updated in ~5 months. A rebase is needed before further review iterations are useful.

**Outstanding review feedback (from Jan 3-5) still unresolved:**

1. **API naming: `PartitionStrategy` vs `Progression`** -- @cheyang suggested renaming `PartitionStrategy` to `Progression` to better convey intent from the user's perspective. This hasn't been addressed yet.

2. **Field/JSON tag mismatch** -- The Go struct field is `SegmentSize` but the JSON tag is `"segment"`. This is confusing for anyone reading the types. Pick one consistent name.

3. **Input validation for segment sizes** -- There is no CRD-level or controller-level validation that segment sizes are positive integers and don't exceed total desired replicas. A zero `segmentSize` currently just gets a `continue` in `calculateMinFullSegment`, but the user gets no feedback that their config is effectively ignored. Add `minimum: 1` in the CRD schema and a webhook/controller-level check that `segmentSize <= role.Replicas`.

4. **Validate role names in segment config exist in RBG spec** -- If a user typos a role name in the segment map, it silently does nothing. Should return a validation error.

**Additional observations on this revision:**

5. **`isSegmentReady` when `segmentIndex == 0`** -- Returning `true` here means "no segments deployed yet, ready to deploy first." The naming is slightly misleading since nothing is actually "ready." Consider renaming to `canAdvancePastSegment` or adding a comment clarifying the semantic.

6. **Test coverage gaps** -- The test file covers the happy paths well but lacks edge-case coverage for: zero segment size, roles in segment config that don't exist in the RBG spec, and scale-down scenarios (current replicas > target). These are the cases most likely to surface bugs in production.

7. **KEP checklist** -- The graduation criteria section is still mostly TODO placeholders. Worth filling in even at alpha stage to set expectations.

---

**Verdict: needs-work.** Please rebase onto current main, address the unresolved API naming and validation issues from the previous review round, and add the missing edge-case tests. Happy to re-review once those are in.

**Comment by cheyang** [comment]
This PR currently has merge conflicts with the `main` branch. Could you please rebase onto the latest `main` so we can move forward with the review? Thanks!
